src/analyzer/protocol/ssl/SSL.cc

```diff
@@ -98,4 +127,279 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
+		const u_char* c_iv = keys.data() + 64;
+		// server IV
+		const u_char* s_iv = keys.data() + 68;
+
+		// FIXME: should we change types here?
+		u_char* encrypted = (u_char*)data;
+		size_t encrypted_len = len;
+
+		if ( is_orig )
+			c_seq++;
+		else
+			s_seq++;
+
+		// AEAD nonce, length 12
+		std::basic_string<unsigned char> s_aead_nonce;
+		if ( is_orig )
+			s_aead_nonce.assign(c_iv, 4);
+		else
+			s_aead_nonce.assign(s_iv, 4);
+
+		// this should be the explicit counter
+		s_aead_nonce.append(encrypted, 8);
+		assert(s_aead_nonce.size() == 12);
+
+		EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
+		EVP_CIPHER_CTX_init(ctx);
+		EVP_CipherInit(ctx, EVP_aes_256_gcm(), NULL, NULL, 0);
+
+		encrypted += 8;
+		// FIXME: is this because of nonce and aead tag?
+		if ( encrypted_len <= (16 + 8) )
+			{
+			DBG_LOG(DBG_ANALYZER, "Invalid encrypted length encountered during TLS decryption");
+			EVP_CIPHER_CTX_free(ctx);
+			return false;
+			}
+		encrypted_len -= 8;
+		encrypted_len -= 16;
+
+		// FIXME: aes_256_gcm should not be hardcoded here ;)
+		if ( is_orig )
+			EVP_DecryptInit(ctx, EVP_aes_256_gcm(), c_wk, s_aead_nonce.data());
+		else
+			EVP_DecryptInit(ctx, EVP_aes_256_gcm(), s_wk, s_aead_nonce.data());
+
+		EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, 16, encrypted + encrypted_len);
+
+		// AEAD tag
+		std::basic_string<unsigned char> s_aead_tag;
+		if ( is_orig )
+			s_aead_tag = fmt_seq(c_seq);
+		else
+			s_aead_tag = fmt_seq(s_seq);
+
+		s_aead_tag[8] = content_type;
+		s_aead_tag[9] = MSB(raw_tls_version);
+		s_aead_tag[10] = LSB(raw_tls_version);
+		s_aead_tag[11] = MSB(encrypted_len);
+		s_aead_tag[12] = LSB(encrypted_len);
+		assert(s_aead_tag.size() == 13);
+
+		auto decrypted = std::vector<u_char>(
+			encrypted_len +
+			16); // see OpenSSL manpage - 16 is the block size for the supported cipher
+		int decrypted_len = 0;
+
+		EVP_DecryptUpdate(ctx, NULL, &decrypted_len, s_aead_tag.data(), s_aead_tag.size());
+		EVP_DecryptUpdate(ctx, decrypted.data(), &decrypted_len, (const u_char*)encrypted,
+		                  encrypted_len);
+		assert(decrypted_len <= decrypted.size());
+		decrypted.resize(decrypted_len);
+
+		int res = 0;
+		if ( ! (res = EVP_DecryptFinal(ctx, NULL, &res)) )
+			{
+			DBG_LOG(DBG_ANALYZER, "Decryption failed with return code: %d. Invalid key?\n", res);
+			EVP_CIPHER_CTX_free(ctx);
+			return false;
+			}
+
+		DBG_LOG(DBG_ANALYZER, "Successfully decrypted %d bytes.", decrypted_len);
+		EVP_CIPHER_CTX_free(ctx);
+		ForwardDecryptedData(decrypted, is_orig);
+
+		return true;
+		}
+
+	// This is only reached if key derivation fails or is unsupported
+	return false;
+	}
+
+void SSL_Analyzer::ForwardDecryptedData(const std::vector<u_char>& data, bool is_orig)
+	{
+	if ( ! pia )
+		{
+		pia = new analyzer::pia::PIA_TCP(Conn());
+		if ( AddChildAnalyzer(pia) )
+			{
+			pia->FirstPacket(true, nullptr);
+			pia->FirstPacket(false, nullptr);
+			}
+		else
+			reporter->FatalError("Could not initialize PIA");
```

Comment:
Let's not use a fatal error here, that would abort Zeek (I realize this shouldn't happen, but who knows what somebody else does to the analyzer tree)

Reply:
Downgrading this to just an error.